### PR TITLE
[SPARK-29844][ML] Improper unpersist strategy in ml.recommendation.ASL.train

### PR DESCRIPTION
**Pull request by Icysandwich**
### What changes were proposed in this pull request?
Adjust improper unpersist timing on RDD.

### Why are the changes needed?
Improper unpersist timing will result in memory waste

### Does this PR introduce any user-facing change?
No

### How was this patch tested?
Manually

### CODE REVIEW

File: mllib/src/main/scala/org/apache/spark/ml/recommendation/ALS.scala [hunk 1]
```diff
@@ -1041,13 +1041,13 @@ object ALS extends DefaultParamsReadable[ALS] with Logging {
...
     if (finalRDDStorageLevel != StorageLevel.NONE) {
       userIdAndFactors.count()
-      itemIdAndFactors.count()
-      itemFactors.unpersist()
       userInBlocks.unpersist()
       userOutBlocks.unpersist()
-      itemInBlocks.unpersist()
       itemOutBlocks.unpersist()
       blockRatings.unpersist()
+      itemIdAndFactors.count()
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/ml/recommendation/ALS.scala]: I don't see the problem here immediately. Doesn't this depend on some of the cached user RDDs?

Comment (Icysandwich) [reply mllib/src/main/scala/org/apache/spark/ml/recommendation/ALS.scala]: itemIdAndFactors only directly depends on itemFactors and itemInBlocks.

Comment (Icysandwich) [reply mllib/src/main/scala/org/apache/spark/ml/recommendation/ALS.scala]: itemIdAndFactors only directly depends on itemFactors and itemInBlocks.

Comment (srowen) [reply mllib/src/main/scala/org/apache/spark/ml/recommendation/ALS.scala]: Yes, but itemFactors depends on the previous user RDDs, including userOutBlocks. I think this is done on purpose.

Comment (Icysandwich) [reply mllib/src/main/scala/org/apache/spark/ml/recommendation/ALS.scala]: Yep, that's my point.

**Review by srowen — APPROVED** [review]

**Comment by srowen** [comment]
Merged to master

**Pull request closed without merging**